src/colmap/scene/frame.h

```
@@ -0,0 +1,337 @@
+
+  // Access the frame from world transformation
+  inline const Rigid3d& FrameFromWorld() const;
+  inline Rigid3d& FrameFromWorld();
+  inline const std::optional<Rigid3d>& MaybeFrameFromWorld() const;
+  inline std::optional<Rigid3d>& MaybeFrameFromWorld();
+  inline void SetFrameFromWorld(const Rigid3d& frame_from_world);
+  inline void SetFrameFromWorld(const std::optional<Rigid3d>& frame_from_world);
+  inline bool HasPose() const;
+  inline void ResetPose();
+
+  // Get the sensor from world transformation
+  inline Rigid3d SensorFromWorld(sensor_t sensor_id) const;
+
+ private:
+  frame_t frame_id_;
+  std::set<data_t> data_ids_;
+
+  // Store the frame_from_world transformation and an optional rig calibration.
+  // If the rig calibration is a nullptr, the frame becomes a single sensor
+  // case, where rig modeling is no longer needed.
+  std::optional<Rigid3d> frame_from_world_;
+  rig_t rig_id_ = kInvalidRigId;
+  std::shared_ptr<class RigCalibration> rig_calibration_ = nullptr;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+// Implementation
+////////////////////////////////////////////////////////////////////////////////
+
+rig_t RigCalibration::RigId() const { return rig_id_; }
+
+void RigCalibration::SetRigId(rig_t rig_id) { rig_id_ = rig_id; }
+
+bool RigCalibration::HasSensor(sensor_t sensor_id) const {
+  return sensor_ids_.find(sensor_id) != sensor_ids_.end();
+}
+
+size_t RigCalibration::NumSensors() const { return sensor_ids_.size(); }
+
+void RigCalibration::AddReferenceSensor(sensor_t ref_sensor_id) {
+  THROW_CHECK(sensor_ids_.empty());  // The reference sensor must be added first
+  ref_sensor_id_ = ref_sensor_id;
+  sensor_ids_.insert(ref_sensor_id);
+}
+
+void RigCalibration::AddSensor(sensor_t sensor_id,
+                               const Rigid3d& sensor_from_rig) {
+  if (NumSensors() == 0)
+    LOG(FATAL_THROW) << "The reference sensor needs to added first before any "
```

Let's throw an exception instead for consistency.